invenio_rdm_migrator/streams/records/transform.py

```diff
@@ -0,0 +1,132 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio-RDM-Migrator is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Invenio RDM migration record transform interfaces."""
+
+from abc import ABC, abstractmethod
+
+from ...transform import Entry, Transform
+
+
+class RDMRecordTransform(Transform):
+    """Base class for record data transformation."""
+
+    @abstractmethod
+    def _parent(self, entry):
+        """Transform the parent record."""
+        pass
+
+    @abstractmethod
+    def _record(self, entry):
+        """Transform the record."""
+        pass
+
+    @abstractmethod
+    def _draft(self, entry):
+        """Transform the draft."""
+        pass
+
+    @abstractmethod
+    def _record_files(self, entry):
+        """Transform the record files."""
+        pass
+
+    @abstractmethod
+    def _draft_files(self, entry):
+        """Transform the draft files."""
+        pass
+
+    def _transform(self, entry):
+        """Transform a single entry."""
+        # the functions receive the full record/data entry
+        # while in most cases the full view is not needed
+        # since this is a low level tool used only by users
+        # with deep system knowledge providing the flexibility
+        # is future proofing and simplifying the interface
+        return {
+            "record": self._record(entry),
+            "draft": self._draft(entry),
+            "parent": self._parent(entry),
+            "record_files": self._record_files(entry),
+            "draft_files": self._draft_files(entry),
+        }
+
+    def run(self, datagen):
+        """Transform and yield one element at a time."""
+        for entry in datagen:
+            yield self._transform(entry)
+
+
+class RDMRecordEntry(Entry):
+    """Transform a single record entry."""
+
+    @abstractmethod
+    def _created(self, entry):
+        """Returns the creation date of the record."""
+        pass
+
+    @abstractmethod
+    def _updated(self, entry):
+        """Returns the update date of the record."""
+        pass
+
+    @abstractmethod
+    def _version_id(self, entry):
+        """Returns the version id of the record."""
+        pass
```

Comment:
comment/shelve: as I was writing the users stream, I felt that these are a bit too fine-grained methods (and a lot of repetition), for a simply mapped attribute (i.e. this always returns `entry.get("<key_name>")`.

Not to be addressed now, but we might need a small refactor to either bundle these up, or come up with something smart.

Reply:
I agree on this. I think we will converge to something parametrized as you say. We might want to wait for more implementations (e.g. CDS) since we will also most likely have to refactor which methods are abstract and which have a default.